dynatrace/tests/test_dynatrace.py

```diff
@@ -0,0 +1,235 @@
+        self.assertEqual(len(topo_instances['components']), 3)
+        # check the second component data and should match
+        component = topo_instances['components'][1]
+        self.assertEqual(component['id'], 'PROCESS_GROUP-647A29F2E878C508')
+        self.assertEqual(component['type'], 'process-group')
+        self.assertEqual(component['data']['identifiers'], ['urn:process-group:/PROCESS_GROUP-647A29F2E878C508'])
+        self.assertEqual(component['data']['entityId'], 'PROCESS_GROUP-647A29F2E878C508')
+        self.assertEqual(component['data']['displayName'], 'Elasticsearch stackstate')
+        self.assertEqual(len(topo_instances['relations']), 14)
+
+    def test_filter_data(self):
+        """
+        Test to check if relationships are removed from data
+        """
+        data = read_data("host_response.json")[0]
+        result = self.check.filter_data(data.copy())
+        self.assertNotIn("fromRelationships", result)
+        self.assertNotIn("toRelationships", result)
+        self.assertNotEqual(data, result)
+
+    def test_filter_data_without_from_to_relationships(self):
+        """
+        Test to check if data remains same as no relationships existed
+        """
+        data = read_data("host_response.json")[0]
+        del data["fromRelationships"]
+        del data["toRelationships"]
+        result = self.check.filter_data(data.copy())
+        self.assertEqual(data, result)
+
+    def test_collect_relations(self):
+        """
+        Test to check if relations are collected properly
+        """
+
+        # TODO this is needed because the topology retains data across tests
+        topology.reset()
+
+        component = read_data("host_response.json")[0]
+        self.check.collect_relations(component, component.get('entityId'))
+
+        topo_instances = topology.get_snapshot(self.check.check_id)
+        self.assertEqual(len(topo_instances['components']), 0)
+        self.assertEqual(len(topo_instances['relations']), 4)
+
+        # since all relations are to this host itself so target id is same
+        relation = topo_instances['relations'][0]
+        self.assertEqual(relation['target_id'], component.get('entityId'))
+        self.assertIn(relation['type'], ['isProcessOf', 'runsOn', 'isSiteOf'])
+
+    def test_check_raise_exception(self):
+        """
+        Test to raise a check exception when collecting components
+        """
+
+        # TODO this is needed because the aggregator retains data across tests
+        aggregator.reset()
+
+        self.check.get_json_response = mock.MagicMock()
+        self.check.get_json_response.side_effect = Exception("Exception occured")
+
+        self.assertRaises(Exception, self.check.check(self.instance))
+        service_checks = aggregator.service_checks("dynatrace")
+        self.assertEqual(len(service_checks), 1)
+        self.assertEqual(service_checks[0].name, self.SERVICE_CHECK_NAME)
+        self.assertEqual(service_checks[0].status, 2)
```

Comment:
I am missing one test: the end-to-end test where we have data from several responses. This perhaps could also have caught the duplicate relationship problem I pointed out.

Reply:
added full e2e smartscape topology test which has duplicate relations and filtered out.